src/main/scala/io/citrine/lolo/bags/Bagger.scala

```diff
@@ -100,191 +115,102 @@ case class Bagger[T: ClassTag](
-    models: ParSeq[Model[T]],
-    featureImportance: Option[Vector[Double]],
-    Nib: Vector[Vector[Int]],
-    trainingData: Seq[(Vector[Any], T)],
-    useJackknife: Boolean,
-    biasModel: Option[Model[Double]] = None,
-    rescale: Double = 1.0,
-    disableBootstrap: Boolean = false
-) extends TrainingResult[T] {
-
-  lazy val NibT: Seq[Vector[Int]] = Nib.transpose
-  lazy val model = new BaggedModel(models, Nib, useJackknife, biasModel, rescale, disableBootstrap)
-  lazy val rep: T = trainingData.find(_._2 != null).get._2
-  lazy val predictedVsActual: Seq[(Vector[Any], T, T)] = trainingData.zip(NibT).flatMap {
-    case ((f, l), nb) =>
-      val oob = if (disableBootstrap) {
-        models.zip(nb)
-      } else {
-        models.zip(nb).filter(_._2 == 0)
-      }
-
-      if (oob.isEmpty || l == null || (l.isInstanceOf[Double] && l.asInstanceOf[Double].isNaN)) {
-        Seq()
-      } else {
-        val predicted = l match {
-          case _: Double => oob.map(_._1.transform(Seq(f)).getExpected().head.asInstanceOf[Double]).sum / oob.size
-          case _         => oob.map(_._1.transform(Seq(f)).getExpected().head).groupBy(identity).maxBy(_._2.size)._1
-        }
-        Seq((f, predicted.asInstanceOf[T], l))
-      }
-  }
-
-  lazy val loss: Double = rep match {
-    case _: Double => RegressionMetrics.RMSE(predictedVsActual.asInstanceOf[Seq[(Vector[Any], Double, Double)]])
-    case _         => ClassificationMetrics.loss(predictedVsActual)
-  }
-
-  /**
-    * Average the influences across the ensemble of models
-    *
-    * @return feature influences as an array of doubles
-    */
-  override def getFeatureImportance(): Option[Vector[Double]] = featureImportance
-
-  override def getModel(): BaggedModel[T] = model
-
-  override def getPredictedVsActual(): Option[Seq[(Vector[Any], T, T)]] = Some(predictedVsActual)
-
-  override def getLoss(): Option[Double] = {
-    if (predictedVsActual.nonEmpty) {
-      Some(loss)
-    } else {
-      None
-    }
-  }
-}
-
 /**
-  * Container holding a parallel sequence of models and the sample counts used to train them
+  * A bagger creates an ensemble of models by training the learner on random samples of the training data
   *
-  * @param models in this bagged model
-  * @param Nib    training sample counts
+  * @param baseLearner learner to train each model in the ensemble
+  * @param numBags number of base models to aggregate (default of -1 sets the number of models to the number of training rows)
+  * @param useJackknife whether to enable jackknife uncertainty estimate
+  * @param disableBootstrap whether to disable bootstrap (useful when `method` implements its own randomization)
+  * @tparam T the type of label data
   */
-class BaggedModel[+T: ClassTag](
-    models: ParSeq[Model[T]],
-    Nib: Vector[Vector[Int]],
-    useJackknife: Boolean,
-    biasModel: Option[Model[Double]] = None,
-    rescale: Double = 1.0,
+case class ClassificationBagger[T](
```

Comment:
I have been debating whether to make this class `ClassificationBager[T] <: Bagger[T]` or `ClassificationBagger <: Bagger[Any]`. 

My initial thought was that since this class is wrapping a `baseLearner[T]`, we should respect and carry through that type of label. But in practice, this is only used to wrap a `ClassificationTree <: Learner[Any]`, so the generic is not strictly necessary to make the code compile.

Reply:
The same consideration applies to my implementation of `ClassificationStandardizer[T] <: Standardizer[T]`.

Reply:
By making it extend `Bagger[T]` you allow for the possibility of someone instantiating, say, `ClassificationBagger[Boolean]` or some other more specific label type. But we don't see any reason to do that. Are there any other practical distinctions between the two choices?

If it's such a small difference, then I don't have a strong opinion either way. I suppose I lean towards extending `Bagger[Any]` based on the principle of "don't create more options than you need." But it's a weak preference.

Reply:
The thought I'm leaning towards as of now is that since this bagger class wraps a `Learner[T]` we should respect the base learner's type, and produce a `TrainingResult[T]`, `Model[T]`, and `PredictionResult[T]`, rather than "erasing" the type into a `Learner[Any]` on instantiation. I also that since `Learner[T]` is invariant in its type parameter, that "erasing" to `Any` would actually not work, although I can test this..

Either way, since the code is actually implemented generically now for `ClassificationBagger[T]`, I still would opt to keep it unless you object otherwise.

Reply:
@sfriedowitz, regarding this. Didn't you say in practice the classification data always comes in as a string?

Reply:
It does from Mithril, at least, where the label type is derived from the associated descriptor value types. This library is a bit more generic.

Reply:
One of the considerations here is that within the `trees` code, categorical is encoded with a function that maps `T => Char`, and the actual classification learning is performed on the low-memory `Char` data type. Some of these `Learner` implementations are required to operate on either top-level `T`-valued data, or low-level `Char`-valued data. So in those cases, the generics are necessary (e.g., `GuessTheModeLearner`).